core/templates/dev/head/pages/exploration_editor/ExplorationSaveService.js

```diff
@@ -0,0 +1,559 @@
+    };
+
+    var areRequiredFieldsFilled = function() {
+      if (!explorationTitleService.displayed) {
+        alertsService.addWarning('Please specify a title');
+        return false;
+      }
+      if (!explorationObjectiveService.displayed) {
+        alertsService.addWarning('Please specify an objective');
+        return false;
+      }
+      if (!explorationCategoryService.displayed) {
+        alertsService.addWarning('Please specify a category');
+        return false;
+      }
+
+      return true;
+    };
+
+    var showCongratulatorySharingModal = function() {
+      return $modal.open({
+        templateUrl: 'modals/shareExplorationAfterPublish',
+        backdrop: true,
+        controller: [
+          '$scope', '$modalInstance', 'explorationContextService',
+          'UrlInterpolationService',
+          function($scope, $modalInstance, explorationContextService,
+            UrlInterpolationService) {
+            $scope.congratsImgUrl = UrlInterpolationService.getStaticImageUrl(
+              '/general/congrats.svg');
+            $scope.DEFAULT_TWITTER_SHARE_MESSAGE_EDITOR = (
+              GLOBALS.DEFAULT_TWITTER_SHARE_MESSAGE_EDITOR);
+            $scope.close = function() {
+              $modalInstance.dismiss('cancel');
+            };
+            $scope.explorationId = (
+              explorationContextService.getExplorationId());
+          }
+        ]
+      });
+    };
+
+    var openPublishExplorationModal = function(
+        onStartSaveCallback, onSaveDoneCallback) {
+      var saveEventCallbacksAreInUse =
+        onStartSaveCallback && onSaveDoneCallback;
+
+      // This is resolved when modal is closed.
+      var whenModalClosed = $q.defer();
+
+      var publishModalInstance = $modal.open({
+        templateUrl: 'modals/publishExploration',
+        backdrop: true,
+        controller: [
+          '$scope', '$modalInstance', function($scope, $modalInstance) {
+            $scope.publish = $modalInstance.close;
+
+            $scope.cancel = function() {
+              $modalInstance.dismiss('cancel');
+              alertsService.clearWarnings();
+              whenModalClosed.resolve();
+            };
+          }
+        ]
+      });
+
+      publishModalInstance.result.then(function() {
+        if (saveEventCallbacksAreInUse) {
+          onStartSaveCallback();
+        };
+
+        explorationRightsService.saveChangeToBackend({
+          is_public: true
+        }).then(function() {
+          if (saveEventCallbacksAreInUse) {
+            onSaveDoneCallback();
+          };
+
+          showCongratulatorySharingModal();
+          siteAnalyticsService.registerPublishExplorationEvent(
+            explorationData.explorationId);
+          whenModalClosed.resolve();
+        });
+      });
+
+      return whenModalClosed.promise;
+    };
+
+    var saveDraftToBackend = function(commitMessage) {
+      // Resolved when save is done.
```

Comment:
Maybe clarify comment: "Resolved when save is done (regardless of success or failure of the operation)."